src/libUtils/DnsUtils.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2021 Zilliqa
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "DnsUtils.h"
+#include "DataConversion.h"
+#include "Logger.h"
+
+#include "common/Constants.h"
+#include "libUtils/DetachedFunction.h"
+#include "libUtils/IPConverter.h"
+
+#include <arpa/inet.h>
+#include <arpa/nameser.h>
+#include <netdb.h>
+#include <resolv.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <condition_variable>
+#include <unordered_map>
+
+using namespace std;
+
+namespace {
+
+atomic<bool> isShuttingDown;
+
+using DnsIpListData = vector<string>;
+using DnsPubKeyListData = unordered_map<uint128_t, bytes>;
+
+struct DnsCacheList {
+  mutex dataAccessMutex;
+
+  DnsIpListData ipList;
+  DnsPubKeyListData pubKeyList;
+};
+
+unordered_map<DnsListType, string> dnsAddresses;
+
+// End of dsepoch, a separate thread will be spawned to query
```

Comment:
I think to be more exact, it's end of DS block consensus or processing, right? Not end of DS epoch